core/priority/calculate.go

```diff
@@ -0,0 +1,158 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package priority
+
+import (
+	"math/rand"
+	"sort"
+
+	"github.com/obolnetwork/charon/app/errors"
+	pbv1 "github.com/obolnetwork/charon/core/corepb/v1"
+)
+
+const (
+	maxPriorities = 1000          // maxPriorities defines the max priorities by topic.
+	countWeight   = maxPriorities // Weight count more than relative priority.
+)
+
+// calculateResults is a deterministic function that returns the cluster wide
+// priorities given the priorities of each peer. Priorities are included in the
+// result if minRequired peers provided them and are ordered by number of peers
+// then by overall priority.
+//
+// Peer messages are validated such that
+// they do not contain duplicate peers, they contain identical slots,
+// individual peers may not contain duplicate topics, individual topics
+// may not contain duplicate priorities or more than 1000 priorities.
+//nolint:gocognit // Mostly just map reduce.
+func calculateResults(msgs []*pbv1.PriorityMsg, minRequired int) (*pbv1.PriorityResult, error) {
```

Comment:
It would be great if you mention the formula that you are using to calculate result in comments

Reply:
Isn't this is?  That is all you need to know I feel
```
Priorities are included in the
// result if minRequired peers provided them and are ordered by number of peers
// then by overall priority.
```